src/guiengine/widgets/ribbon_widget.cpp

```diff
@@ -666,6 +647,51 @@ EventPropagation RibbonWidget::propagationType(const bool horizontally)
     return result;
 }
 
+/**
+ * Move to the next child widget in the requested direction.
+ * If it is inactive, move again, until it finds an active child or go over 100 loops
+ */
+void RibbonWidget::selectNextActiveWidget(const bool horizontally, const bool reverse,
+                                          const int playerID, const int old_selection)
+{
+    int loop_counter = 0;
+    do
+    {
+        if (reverse)
+            m_selection[playerID]--;
+        else
+            m_selection[playerID]++;
+
+        if (m_selection[playerID] >= int(m_active_children.size()) || m_selection[playerID] < 0)
+        {
+            // In vertical tabs, don't loop when reaching the top or bottom
+            if (!horizontally)
+            {
+                if (reverse)
+                    m_selection[playerID] = old_selection;
+                else
+                    m_selection[playerID] = old_selection;
+
+                return;
+            }
+            bool left = (m_selection[playerID] < 0);
+
+            if (m_listener != NULL) m_listener->onRibbonWidgetScroll(left ? -1 : 1);
+
+            bool select_zero = (m_event_handler && left) || (!m_event_handler && !left);
+
+            m_selection[playerID] = select_zero ? 0 : m_active_children.size()-1;
+        }
+
+        loop_counter++;
+        if (loop_counter > 100)
+        {
+            Log::warn("RibbonWidget", "Empty ribbon widget focused !");
```

Comment:
Can't you just detect that the ribbon is empty by checking for the number of children, instead of counting loop iterations?

Reply:
The list of children still keeps deactived children afaict. That's used when a ribbon disable some children under some condition, but enable them in other (see the GP editor). Loop counting is somewhat necessary.

Though rather than an arbitrary 100, I think it would work with m_active_children.size() or something like that.